### PR TITLE
remove atomic arc and mutex since the monoio is based on thread per c…

### PR DESCRIPTION
**Pull request by rainj-me**
…ore architecture

### CODE REVIEW

File: monoio-http/src/h2/proto/ping_pong.rs [hunk 1]
```diff
@@ -215,20 +211,15 @@ impl ReceivedPing {
...
+        let user_pings = unsafe { &mut *self.0.get() };
+        let prev = user_pings.state; // current
+        if prev == USER_STATE_EMPTY {
+            user_pings.state = USER_STATE_PENDING_PING; // new
+        }
 
         match prev {
             USER_STATE_EMPTY => {
-                self.0.ping_task.wake();
+                user_pings.ping_task.take().unwrap().wake();
```
Comment (ihciah) [inline monoio-http/src/h2/proto/ping_pong.rs:222]: Here we should create a new `waker.rs` like what atomic-waker does.
1. Register waker should compare waker like `old_waker.will_wake(waker)`, if these two wakers not equal, then do waker.clone(), replace and wake old waker.
2. `wake` should not panic when inner waker is None. Use if let.

File: monoio-http/src/h2/proto/ping_pong.rs [hunk 2]
```diff
@@ -242,17 +233,12 @@ impl UserPings {
...
-            .state
-            .compare_exchange(
-                USER_STATE_RECEIVED_PONG, // current
-                USER_STATE_EMPTY,         // new
-                Ordering::AcqRel,
-                Ordering::Acquire,
-            )
-            .unwrap_or_else(|v| v);
+        let user_pings = unsafe { &mut *self.0.get() };
+        user_pings.pong_task = Some(cx.waker().clone());
```
Comment (ihciah) [inline monoio-http/src/h2/proto/ping_pong.rs:237]: Same problem. We don't have to clone waker if equal.

File: monoio-http/src/h2/proto/streams/streams.rs [hunk 3]
```diff
@@ -184,16 +180,14 @@ where
     where
         T: AsyncWriteRent + Unpin + 'static,
     {
-        let mut me = self.inner.lock().unwrap();
-        me.poll_complete(&self.send_buffer, cx, dst)
+        let me = unsafe { &mut *self.inner.get() };
+        me.poll_complete(self.send_buffer.clone(), cx, dst)
```
Comment (ihciah) [inline monoio-http/src/h2/proto/streams/streams.rs:184]: Maybe don't have to clone?

File: monoio-http/src/h2/proto/streams/streams.rs [hunk 4]
```diff
@@ -619,13 +602,12 @@ impl Inner {
 
     fn recv_window_update<B>(
         &mut self,
-        send_buffer: &SendBuffer<B>,
+        send_buffer: Rc<UnsafeCell<SendBuffer<B>>>,
```
Comment (ihciah) [inline monoio-http/src/h2/proto/streams/streams.rs:605]: Keep & should be ok.

File: monoio-http/src/h2/proto/streams/streams.rs [hunk 5]
```diff
@@ -656,11 +638,14 @@ impl Inner {
         Ok(())
     }
 
-    fn handle_error<B>(&mut self, send_buffer: &SendBuffer<B>, err: proto::Error) -> StreamId {
+    fn handle_error<B>(
+        &mut self,
+        send_buffer: Rc<UnsafeCell<SendBuffer<B>>>,
```
Comment (ihciah) [inline monoio-http/src/h2/proto/streams/streams.rs:643]: Some problem. Keep using & and get `&SendBuffer<B>` at the caller.

File: monoio-http/src/h2/proto/streams/streams.rs [hunk 6]
```diff
@@ -1413,25 +1358,14 @@ impl Clone for OpaqueStreamRef {
 
 impl Drop for OpaqueStreamRef {
     fn drop(&mut self) {
-        drop_stream_ref(&self.inner, self.key);
+        drop_stream_ref(self.inner.clone(), self.key);
```
Comment (ihciah) [inline monoio-http/src/h2/proto/streams/streams.rs:1361]: Don't clone.

**Review by ihciah — CHANGES_REQUESTED** [review]
Overall no problem. But there're some small problems.
1. We should define a local-threaded version atomic-waker, and the action should be alike.
2. For functions with parameters like `&XXX`, we can keep it and do borrow at the caller.
3. There's no need to add new clone.

**Comment by ihciah** [comment]
I only comment on part of them. You can check if there's any other places with the alike issues.

File: monoio-http/src/h2/proto/ping_pong.rs (hunk 1 above)
Comment (rainj-me) [reply monoio-http/src/h2/proto/ping_pong.rs]: fixed

File: monoio-http/src/h2/proto/streams/streams.rs [hunk 7]
```diff
@@ -106,7 +107,7 @@ struct Actions {
 /// Contains the buffer of frames to be written to the wire.
 #[derive(Debug)]
 struct SendBuffer<B> {
-    inner: Mutex<Buffer<Frame<B>>>,
+    inner: Buffer<Frame<B>>,
 }
```
Comment (ihciah) [inline monoio-http/src/h2/proto/streams/streams.rs:111]: Put the UnsafeCell at the place of Mutex is better. It’s good for later cherry pick changes.

**Pull request merged**